src/test/java/com/google/neighborgood/UserInfoServletTest.java

```diff
@@ -0,0 +1,422 @@
+
+    // Filter out the entity that has the userId of 1234567890
+    PreparedQuery results =
+        ds.prepare(
+            new Query("UserInfo")
+                .setFilter(new FilterPredicate("userId", FilterOperator.EQUAL, "1234567890")));
+    Entity entity = results.asSingleEntity();
+    if (entity == null) {
+      // The entity can't be null
+      assertTrue(false);
+    }
+
+    // Test the stored personal information
+    assertEquals("Leonard", (String) entity.getProperty("nickname"));
+    assertEquals("4xxx Centre Avenue", (String) entity.getProperty("address"));
+    assertEquals("4xxxxxxxxx", (String) entity.getProperty("phone"));
+    assertEquals("leonardzhang@google.com", (String) entity.getProperty("email"));
+    assertEquals("1234567890", (String) entity.getProperty("userId"));
+    assertEquals(0, (long) entity.getProperty("points"));
+
+    when(request.getParameter("nickname-input")).thenReturn("Leo");
+
+    try {
+      new UserInfoServlet().doPost(request, response);
+    } catch (IOException e) {
+      assertTrue(false);
+    }
+
+    // After sending the second POST request, there should be still three entities in the datastore
+    assertEquals(3, ds.prepare(new Query("UserInfo")).countEntities(withLimit(10)));
+    results =
+        ds.prepare(
+            new Query("UserInfo")
+                .setFilter(new FilterPredicate("userId", FilterOperator.EQUAL, "1234567890")));
+    entity = results.asSingleEntity();
+    if (entity == null) {
+      // The entity can't be null
+      assertTrue(false);
+    }
+
+    // Test the stored personal information
+    assertEquals("Leo", (String) entity.getProperty("nickname"));
+    assertEquals("4xxx Centre Avenue", (String) entity.getProperty("address"));
+    assertEquals("4xxxxxxxxx", (String) entity.getProperty("phone"));
+    assertEquals("leonardzhang@google.com", (String) entity.getProperty("email"));
+    assertEquals("1234567890", (String) entity.getProperty("userId"));
+    assertEquals(0, (long) entity.getProperty("points"));
+  }
+
+  @Test
+  public void edgeCaseTest() {
+    HttpServletRequest request = mock(HttpServletRequest.class);
+    HttpServletResponse response = mock(HttpServletResponse.class);
+
+    // Check whether the datastore is empty before the test
+    DatastoreService ds = DatastoreServiceFactory.getDatastoreService();
+    assertEquals(0, ds.prepare(new Query("UserInfo")).countEntities(withLimit(10)));
+
+    // Set the nickname input to be empty
+    when(request.getParameter("nickname-input")).thenReturn("");
+    when(request.getParameter("address-input")).thenReturn("4xxx Centre Avenue");
+    when(request.getParameter("phone-input")).thenReturn("4xxxxxxxxx");
+
+    try {
+      new UserInfoServlet().doPost(request, response);
+    } catch (IOException e) {
+      assertTrue(false);
+    }
+
+    // The POST request should lead to an error handling clause, there should be no entity in the
+    // datastore
```

Comment:
It would be good to assert something about the response, as well, such as the status code that is send back to the browser when the input is bad.

Reply:
For my current implementation, the error handling clause will print an error message and redirect the user to a 400 error page. So I think I would test the system output here since it's hard to test the page redirect.

Reply:
Sounds good. Can you also assert something about the status code itself, e.g. using `response.getStatus()`?